hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+    class dist_object
+      : hpx::components::client_base<dist_object<T>,
+            server::dist_object_part<T>>
+    {
+        typedef hpx::components::client_base<dist_object<T>,
+            server::dist_object_part<T>>
+            base_type;
+
+        typedef typename server::dist_object_part<T>::data_type data_type;
+
+    private:
+        template <typename Arg>
+        static hpx::future<hpx::id_type> create_server(Arg&& value)
+        {
+            return hpx::local_new<server::dist_object_part<T>>(
+                std::forward<Arg>(value));
+        }
+
+    public:
+        /// Creates a dist_object in every locality
+        ///
+        /// A dist_object \a base_name is created through default constructor.
+        dist_object() {}
+
+        /// Creates a dist_object in every locality with a given base_name string,
+        /// data, and a type and construction_type in the template parameters
+        ///
+        /// \param construction_type The construction_type in the template parameters
+        /// accepts either Meta_Object, and it is set to All_to_All by defalut
+        /// The Meta_Object option provides meta object registration in the root
+        /// locality and meta object is essentailly a table that can find the
+        /// instances of dist_object in all localities. The All_to_All option only
+        /// locally holds the client and server of the dist_object.
+        /// \param base_name The name of the dist_object, which should be a unique
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type const& data)
+          : base_type(create_server(data))
+          , base_(base)
+        {
+            HPX_ASSERT(C == construction_type::All_to_All ||
+                C == construction_type::Meta_Object);
+            if (C == construction_type::Meta_Object)
+            {
+                meta_object mo(base);
+                basename_list = mo.registration(this->get_id());
+                basename_registration_helper(base);
+            }
+            else
+            {
+                basename_registration_helper(base);
+            }
+        }
+
+        /// Creates a dist_object in every locality with a given base_name string,
+        /// data. The construction_type in the template parameter is set to
+        /// All_to_All option by default.
+        ///
+        /// \param base_name The name of the dist_object, which should be a unique
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type&& data)
+          : base_type(create_server(std::move(data)))
+          , base_(base)
+        {
+            basename_registration_helper(base);
+        }
+        /// \cond NOINTERNAL
+        dist_object(hpx::future<hpx::id_type>&& id)
```

Comment:
what's the semantic of this constructor?

Reply:
If I remeber correctly from @hkaiser, this line generates boilerplate code for the component client. So no error will occur.

Reply:
well, I think this is a design error. Is a `distributed_object` ever to be constructed from a `hpx::future<hpx::id_type>&& id`, what does this mean? I think this is a similar problem as with the serialize function.

Reply:
This might be needed for `distributed_object<>::base_type::then` to compile. We could make the base class a `friend` and hide the constructor as being `private`.

Reply:
Does `distributed_object<>::base_type::then` make any sense? The construction of a `distributed_object` should be almost immediate, no?
I think there are more functions from `hpx::components::client_base` that don't make sense for `distributed_object`: All assignment operators taking a `hpx::id_type` or `hpx::future<hpx::id_type>`, `create` and `create_colocated`, `register_as`, `connect_to`, `reset_registered_name`. I think `hpx::components::client_base` is the wrong abstraction for `distributed_object`.

Reply:
`register_as` is definitely needed as siblings on other localities connect through the symbol namespace if needed. But I concur to your point, `client_base` is too heavy for this specific use case.